src/ReducedModelControllers/src/CentroidalMPC.cpp

```diff
@@ -0,0 +1,1267 @@
+
+            for (const auto& corner : contact.corners)
+            {
+                odeInput.push_back(corner.force);
+            }
+        }
+
+        // set the feedback
+        this->opti.subject_to(this->optiVariables.comCurrent == com(Sl(), 0));
+        this->opti.subject_to(this->optiVariables.dcomCurrent == dcom(Sl(), 0));
+        this->opti.subject_to(this->optiVariables.angularMomentumCurrent
+                              == angularMomentum(Sl(), 0));
+        for (const auto& [key, contact] : this->optiVariables.contacts)
+        {
+            this->opti.subject_to(this->optiVariables.contacts.at(key).currentPosition
+                                  == contact.position(Sl(), 0));
+        }
+
+        // set the dynamics
+        // map computes the multiple shooting method
+        auto dynamics = this->ode().map(this->optiSettings.horizon);
+        auto fullTrajectory = dynamics(odeInput);
+        this->opti.subject_to(extractFutureValuesFromState(com) == fullTrajectory[0]);
+        this->opti.subject_to(extractFutureValuesFromState(dcom) == fullTrajectory[1]);
+        this->opti.subject_to(extractFutureValuesFromState(angularMomentum) == fullTrajectory[2]);
+
+        // footstep dynamics
+        std::size_t contactIndex = 0;
+        for (const auto& [key, contact] : this->optiVariables.contacts)
+        {
+            this->opti.subject_to(extractFutureValuesFromState(contact.position)
+                                  == fullTrajectory[3 + contactIndex]);
+            contactIndex++;
+        }
+
+        // add constraints for the contacts
+        auto contactPositionErrorMap = this->contactPositionError().map(this->optiSettings.horizon);
+
+        // convert the eigen matrix into casadi
+        // please check https://github.com/casadi/casadi/issues/2563 and
+        // https://groups.google.com/forum/#!topic/casadi-users/npPcKItdLN8
+        // Assumption: the matrices are stored as column-major
+        casadi::DM frictionConeMatrix = casadi::DM::zeros(frictionCone.getA().rows(), //
+                                                          frictionCone.getA().cols());
+
+        std::memcpy(frictionConeMatrix.ptr(),
+                    frictionCone.getA().data(),
+                    sizeof(double) * frictionCone.getA().rows() * frictionCone.getA().cols());
+
+        const casadi::DM zero = casadi::DM::zeros(frictionCone.getA().rows(), 1);
+        casadi::MX rotatedFrictionCone;
+
+        for (const auto& [key, contact] : this->optiVariables.contacts)
+        {
+            auto error
+                = contactPositionErrorMap({extractFutureValuesFromState(contact.position),
+                                           extractFutureValuesFromState(contact.nominalPosition),
+                                           contact.orientation});
+
+            this->opti.subject_to(contact.lowerLimitPosition <= error[0]
+                                  <= contact.upperLimitPosition);
```

Comment:
Thank you! Done here: 2fd5622d65ab052627ec6209e03077dcdd466b61

Reply:
I had to vectorize the matrices Since elementwise (in)equalities for matrices are not supported with a natural syntax since there is an ambiguity with semi-definiteness constraints. See https://web.casadi.org/docs/#problem-specification 5534c3caa

Reply:
I've just noticed that splitting the constraint as you suggested increase the number of iteration 🤔 

### Without splitting
```
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       nlp_f  | 199.00us ( 10.47us) 201.92us ( 10.63us)        19
       nlp_g  | 594.00us ( 31.26us) 593.17us ( 31.22us)        19
  nlp_grad_f  | 473.00us ( 23.65us) 475.31us ( 23.77us)        20
  nlp_hess_l  | 690.00us ( 38.33us) 681.59us ( 37.87us)        18
   nlp_jac_g  |   1.10ms ( 54.80us)   1.12ms ( 55.79us)        20
       total  |  33.84ms ( 33.84ms)  33.84ms ( 33.84ms)         1
```

### Splitting
```
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       nlp_f  | 360.00us ( 10.00us) 367.10us ( 10.20us)        36
       nlp_g  |   1.19ms ( 32.94us)   1.20ms ( 33.37us)        36
  nlp_grad_f  | 590.00us ( 19.03us) 591.45us ( 19.08us)        31
  nlp_hess_l  |   1.14ms ( 39.17us)   1.14ms ( 39.26us)        29
   nlp_jac_g  |   1.62ms ( 52.42us)   1.63ms ( 52.50us)        31
       total  |  55.10ms ( 55.10ms)  55.10ms ( 55.10ms)         1
```

It's worth understanding why 🤔 
For the time being, I will restore the original version of the code